[MetaSchedule] Enhance Conv2d NCHW Winograd Schedule Rules

zxybazh opened the pull request:
This PR introduces some new trials to expand the search space for Winograd Conv2d Workload by writing block level schedule rules. Tests will also be included.

zxybazh commented:
@tvm-bot rerun

vinx13 commented on src/meta_schedule/schedule_rule/winograd.cc:
```diff
@@ -92,6 +124,50 @@ TVM_REGISTER_GLOBAL("meta_schedule.winograd_data_pack.llvm")
+                     /*preserve_unit_loops=*/true);
+      if (OL.defined()) {
+        while (sch->GetConsumers(OL).size() > 0) {
+          BlockRV next_OL = GetOnlyConsumer(sch, OL);
```
Will consumers be handled automatically by auto_inline rule? If so, this and `SetScope` are not necessary

zxybazh commented on src/meta_schedule/schedule_rule/winograd.cc:
```diff
@@ -74,6 +80,32 @@ inline LoopRV ScheduleDataPack(Schedule sch, BlockRV block) {
+
+  Array<ExprRV> factors = sch->SamplePerfectTile(loops[3], /*n=*/2, /*max_innermost_factor=*/64);
+  Array<LoopRV> split =
+      sch->Split(loops[3], /*factors=*/{factors[0], factors[1]}, /*preserve_unit_loops=*/true);
```
Hi, it seems `factors` are type of `Array<PrimExpr>` while the argument here requires `factors` to be `Array<Optional<PrimExpr>>` so IIUC we may keep it this way.

zxybazh commented:
Hi Wuwei, I have removed bgemm block schedule and use MultiLevelTiling instead. It reproduced same performance in conv2d test, I’m running e2e tests as well. With other issues also addressed, would you please take another look? Thanks!

vinx13 commented on tests/python/unittest/test_meta_schedule_space_cuda.py:
```diff
@@ -1239,6 +1239,90 @@ def cbr_0(data: T.Buffer[(1, 224, 224, 3), "float32"], kernel: T.Buffer[(7, 7, 3
     )
 
 
+def test_cuda_tbg():
```
this test case is not related

zxybazh replied:
This test case was moved to the right position before the PR, it got moved after the conv2d winograd testcase because of rebase. You can see there’s no change in the `files changed` tab page.

vinx13 commented on tests/python/unittest/test_meta_schedule_space_cuda.py:
```diff
@@ -1378,6 +1456,9 @@ def winograd_nchw_conv2d(data: T.Buffer[(1, 64, 224, 224), "float32"], kernel: T
         sch_rules="default",
     ).generate_design_space()
+    sch = actual[0]
+    print(sch.mod.script())
+    print(sch.trace)
```
remove it
